pkg/install/install.go

```diff
@@ -46,42 +51,191 @@ var kindToResource = map[string]string{
+			return false, errors.Wrap(err, "error waiting for backupstoragelocations to be ready")
 		}
 
-		c, err := factory.ClientForGroupVersionResource(gvk.GroupVersion(), apiResource, r.GetNamespace())
-		if err != nil {
-			return errors.Wrapf(err, "Error creating client for resource %s", id)
+		bsl := new(apiextv1beta1.CustomResourceDefinition)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredBSL.Object, bsl); err != nil {
+			return false, errors.Wrap(err, "error converting backupstoragelocation from unstructured")
 		}
 
-		_, err = c.Create(&r)
-		if apierrors.IsAlreadyExists(err) {
-			log("already exists, proceeding")
-			continue
+		unstructuredVSL, err := c.Get("volumesnapshotlocations.velero.io", metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		} else if err != nil {
+			return false, errors.Wrap(err, "error waiting for volumesnapshotlocations to be ready")
+		}
+
+		vsl := new(apiextv1beta1.CustomResourceDefinition)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredVSL.Object, vsl); err != nil {
+			return false, errors.Wrap(err, "error converting volumesnapshotlocation from unstructured")
+		}
+
+		isReady = (crdIsReady(*bsl) && crdIsReady(*vsl))
+
+		return true, nil
+	})
+	return isReady, err
+}
+
+func deploymentIsReady(factory client.DynamicFactory) (bool, error) {
+	gvk := schema.FromAPIVersionAndKind(appsv1beta1.SchemeGroupVersion.String(), "Deployment")
+	apiResource := metav1.APIResource{
+		Name:       "deployments",
+		Namespaced: true,
+	}
+	c, err := factory.ClientForGroupVersionResource(gvk.GroupVersion(), apiResource, "velero")
+	if err != nil {
+		return false, errors.Wrapf(err, "Error creating client for deployment polling")
+	}
+	// declare this variable out of scope so we can return it
+	var isReady bool
+	err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+		unstructuredDeployment, err := c.Get("velero", metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return false, nil
 		} else if err != nil {
-			return errors.Wrapf(err, "Error creating resource %s", id)
+			return false, errors.Wrap(err, "error waiting for deployment to be ready")
+		}
+
+		deploy := new(appsv1beta1.Deployment)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredDeployment.Object, deploy); err != nil {
+			return false, errors.Wrap(err, "error converting deployment from unstructured")
+		}
+
+		for _, cond := range deploy.Status.Conditions {
+			if cond.Type == appsv1beta1.DeploymentAvailable {
+				isReady = true
+			}
 		}
+		return true, nil
+	})
+	return isReady, err
+}
+
+// GroupResources groups resources into ResourcesGroups based on whether the resources are CustomResourceDefinitions or other types of kubernetes objects
+// This is useful to wait for readiness before creating CRD objects
+func GroupResources(resources *unstructured.UnstructuredList) []*ResourceGroup {
```

Comment:
@skriss The logic in this for loop appears to work with the print statements inside the loop, but outside the loop, the `crdObjs` slice is only comprised of the `Deployment/velero` object. That's also the last element seen from `resources.Items`.

My suspicion is that this is a variable shadowing issue, but I'm not seeing where that might be occurring.

Reply:
Guessing it's from this line:
https://github.com/heptio/velero/pull/1287/commits/9ae653d712dfdb2974011811b4407d0960c4eba7#diff-90259a58ac8d65a2b486198e4eebb9c3R165

And this line:
https://github.com/heptio/velero/pull/1287/commits/9ae653d712dfdb2974011811b4407d0960c4eba7#diff-90259a58ac8d65a2b486198e4eebb9c3R174

Taking the address of the range variable (`r` in this case) is always a recipe for bugs like this. Easiest thing to do is switch your loop to `for i, r := resource.Items {`, and then in your appends, instead of `&r`, append `resources.Items[i]`. 

The problem is that the memory allocated to `r` is used over and over again; so you're adding pointers to the same memory address over and over, which at the end of the loop will only have the last thing that was in that memory.

Reply:
Yep, that appears to have been it! Thanks.